File: android/measure/src/test/java/sh/measure/android/utils/TestClock.kt
package sh.measure.android.utils

import java.time.Duration
import java.util.concurrent.TimeUnit

internal class TestClock private constructor(private var currentEpochMillis: Long) :
    SystemClock {
    fun setTime(time: Long) {
        this.currentEpochMillis = time
    }

    fun advance(duration: Duration) {
        advance(duration.toMillis(), TimeUnit.MILLISECONDS)
    }

    override fun epochTime(): Long {
        return currentEpochMillis
    }

    override fun monotonicTimeSinceBoot(): Long {
        return currentEpochMillis
    }

    private fun advance(duration: Long, unit: TimeUnit) {
        currentEpochMillis += unit.toMillis(duration)
    }

    companion object {
        // Default time set to Wed Oct 25 2023 18:20:15 GMT+0530
        fun create(timeInMillis: Long = 1698238215): TestClock {
            return TestClock(timeInMillis)
        }
    }
}

File: flutter/packages/measure_flutter/test/utils/test_clock.dart
import 'package:measure_flutter/src/time/system_clock.dart';

class TestClock extends SystemClock {
  int _currentEpochMillis;

  TestClock._(this._currentEpochMillis);

  void setTime(int time) {
    _currentEpochMillis = time;
  }

  void advance(Duration duration) {
    _advanceByMillis(duration.inMilliseconds);
  }

  @override
  int epochTime() {
    return _currentEpochMillis;
  }

  void _advanceByMillis(int milliseconds) {
    _currentEpochMillis += milliseconds;
  }

  /// Default time set to Wed Oct 25 2023 18:20:15 GMT+0530
  static TestClock create([int timeInMillis = 1698238215]) {
    return TestClock._(timeInMillis);
  }
}
